### DISCUSSION TITLE
Moving from array of single subdocument to mixed array - what am I missing?

### DISCUSSION THREAD

**Discussion by AdamGerthel** [Automattic/mongoose#13385] [Q&A] [answered]
Basically, in TypeScript-lingo, I'm changing from `SubDocumentA[]` to `(SubDocumentA | string)[]`. After reading around and checking out the options on how this can be achieved, I've come to the conclusion that an array with mixed type is most suitable because of how deep and fairly complex the complete schema is. More specifically, I'm following the advice from [this discussion](https://github.com/Automattic/mongoose/discussions/12154) about using `hydrate()`. My question is: What do I need to do to achieve the same or similar result in regards to validation etc. for my sub-document when I'm using the "mixed" approach?

Here's my current approach. `node.options` is what was previously an array of the sub-document schema "option" but can now also be a string. For context, I'm using FeathersJS and this is a hook for a service (but that is beside the point).

```typescript
const hydrateOptions: Hook = async (context: HookContext<IScene>) => {
  context.data?.nodes?.forEach(node => {
    node.options = node.options.map(option => {
      if (typeof option === 'object') {
        // Here's where the relevant part starts.
        const hydratedOption = models.SceneOption.hydrate(option)
        const error = hydratedOption.validateSync()

        if (error) {
          throw new Error(error)
        }

        return hydratedOption
        // Here's where the revelant part ends.
      }

      return option
    })
  })
}
```
So - as you can see, I'm hydrating the "options" and validating them. This works 100% as expected. So far so good. **However** - if I submit the same document *without* the `_id` property, it will save without creating a new `_id`. This was not the case before I switched to the mixed approach. My initial thinking was that I could do something simple like this:
```typescript
        if (!hydratedOption._id) {
          hydratedOption._id = new mongoose.Types.ObjectId()
        }
```
But each "option" actually also contains other sub-documents and this happens there too. In short: This approach seems to no longer automatically create missing IDs. 

What can I do to make sure it behaves as, or as close to, the way it was when it was an array of a sub-document schema rather than a mixed array?

PS. I'm using v5 of Mongoose and am currently not able to upgrade to 6 or 7.

**Comment by vkarpov15** [comment, marked as answer]
You could try `const hydratedOption = new models.SceneOption(option); hydratedOption.validateSync();` That will also apply defaults, etc.

**Reply by AdamGerthel** [reply]
Thanks! I just did and it's working as expected.
